File: src/util/main/sp/it/util/collections/list/PrefList.java
package sp.it.util.collections.list;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;

/**
 * List with a preferred item. The item is always contained within list.
 */
public class PrefList<E> extends ArrayList<E> {
	private E pref;

	public PrefList() {
		this(List.of(), null);
	}

	public PrefList(List<E> elements, E pref) {
		addAll(elements);
		setPreferred(pref);
	}

	/**
	 * Get preferred item.
	 *
	 * @return preferred item or null if none
	 */
	public E getPreferred() {
		return pref;
	}

	/**
	 * Get preferred item. First item if no preferred is set.
	 *
	 * @return preferred item or first if none or null list empty
	 */
	public E getPreferredOrFirst() {
		return pref!=null ? pref : isEmpty() ? null : get(0);
	}

	public void addPreferred(E e) {
		pref = e;
		super.add(e);
	}

	public void addPreferred(int index, E element) {
		pref = element;
		super.add(index, element);
	}

	public void addPreferred(E e, boolean preferred) {
		if (preferred) addPreferred(e);
		else add(e);
	}

	public void setPreferred(E e) {
		if (e==null) {
			pref = null;
		} else {
			var i = indexOf(e);
			if (i>=0) pref = get(i);
		}
	}

	@Override
	public boolean remove(Object o) {
		if (pref!=null && pref.equals(o)) pref = null;
		return super.remove(o);
	}

	@Override
	public E remove(int i) {
		if (pref!=null && pref.equals(get(i))) pref = null;
		return super.remove(i);
	}

	@Override
	public boolean removeIf(Predicate<? super E> filter) {
		if (pref!=null && filter.test(pref)) pref = null;
		return super.removeIf(filter);
	}

	@Override
	public boolean removeAll(Collection<?> c) {
		if (pref!=null && c.contains(pref)) pref = null;
		return super.removeAll(c);
	}

	@Override
	public void replaceAll(UnaryOperator<E> operator) {
		int i = pref==null ? -1 : indexOf(pref);
		super.replaceAll(operator);
		if (i!=-1) pref = get(i);
	}

	@Override
	public boolean retainAll(Collection<?> c) {
		if (pref!=null && !c.contains(pref)) pref = null;
		return super.retainAll(c);
	}

	@Override
	public void clear() {
		pref = null;
		super.clear();
	}

	@Override
	protected void removeRange(int fromIndex, int toIndex) {
		if (pref!=null) {
			int i = indexOf(pref);
			if (i>=fromIndex && i<toIndex) pref = null;
		}
		super.removeRange(fromIndex, toIndex);
	}

	public <T> PrefList<T> map(Function<? super E, ? extends T> mapper) {
		var l = new PrefList<T>();
		forEach(it -> {
			if (pref==it) l.addPreferred(mapper.apply(it));
			else l.add(mapper.apply(it));
		});
		return l;
	}
}

File: src/util/main/sp/it/util/collections/list/PrefList.kt
package sp.it.util.collections.list

import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.function.UnaryOperator

/** List with a preferred item. The item is always contained within list. */
class PrefList<E: Any> @JvmOverloads constructor(elements: List<E> = listOf(), preferred: E? = null): ArrayList<E>() {

   private var pref: E? = null

   /** Preferred item or null if none preferred or list empty */
   var preferred: E?
      get() = pref
      set(e) {
         pref = e?.let { if (it in this) it else null }
      }

   /** Preferred item or first item or null if none preferred or list empty */
   val preferredOrFirst: E?
      get() = pref ?: firstOrNull()

   infix fun has(element: E?): Boolean =
      element != null && element in this

   init {
      this += elements
      this.preferred = preferred
   }

   fun addPreferred(e: E, preferred: Boolean = true) {
      if (preferred) {
         pref = e
         super.add(e)
      } else {
         super.add(e)
      }
   }

   override fun remove(element: E): Boolean {
      if (pref == element) pref = null
      return super.remove(element)
   }

   override fun removeAt(at: Int): E {
      if (pref == get(at)) pref = null
      return super.removeAt(at)
   }

   override fun removeIf(filter: Predicate<in E>): Boolean {
      if (pref != null && filter.test(pref!!)) pref = null
      return super.removeIf(filter)
   }

   override fun removeAll(elements: Collection<E>): Boolean {
      if (pref != null && elements.contains(pref)) pref = null
      return super.removeAll(elements)
   }

   override fun replaceAll(operator: UnaryOperator<E>) {
      val i = if (pref == null) -1 else indexOf(pref)
      super.replaceAll(operator)
      if (i!=-1) pref = get(i)
   }

   override fun retainAll(elements: Collection<E>): Boolean {
      if (pref != null && !elements.contains(pref)) pref = null
      return super.retainAll(elements)
   }

   override fun clear() {
      pref = null
      super.clear()
   }

   override protected fun removeRange(fromIndex: Int, toIndex: Int) {
      if (pref != null) {
         val i = indexOf(pref)
         if (i in fromIndex until toIndex) pref = null
      }
      super.removeRange(fromIndex, toIndex)
   }

   fun <T: Any> map(mapper: (E) -> T): PrefList<T> {
      val i = pref?.let { indexOf(it) } ?: -1
      val l = PrefList<T>()
      forEach { l += mapper(it) }
      if (i!=-1) l.preferred = l.get(i)
      return l
   }
}